src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/CSharp/DefaultCSharpCodeActionResolver.cs

```diff
@@ -0,0 +1,165 @@
+            }
+
+            if (documentResolver is null)
+            {
+                throw new ArgumentNullException(nameof(documentResolver));
+            }
+
+            if (razorFormattingService is null)
+            {
+                throw new ArgumentNullException(nameof(razorFormattingService));
+            }
+
+            _foregroundDispatcher = foregroundDispatcher;
+            _documentResolver = documentResolver;
+            _razorFormattingService = razorFormattingService;
+        }
+
+        public override string Action => LanguageServerConstants.CodeActions.Default;
+
+        public async override Task<RazorCodeAction> ResolveAsync(
+            CSharpCodeActionParams csharpParams,
+            RazorCodeAction codeAction,
+            CancellationToken cancellationToken)
+        {
+            if (csharpParams is null)
+            {
+                throw new ArgumentNullException(nameof(csharpParams));
+            }
+
+            if (codeAction is null)
+            {
+                throw new ArgumentNullException(nameof(codeAction));
+            }
+
+            var resolvedCodeAction = await ResolveCodeActionWithServerAsync(codeAction, cancellationToken);
+            if (resolvedCodeAction.Edit?.DocumentChanges is null)
+            {
+                // Unable to resolve code action with server, return original code action
+                return codeAction;
+            }
+
+            if (resolvedCodeAction.Edit.DocumentChanges.Count() != 1)
+            {
+                // We don't yet support multi-document code actions, return original code action
+                return codeAction;
+            }
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var documentSnapshot = await Task.Factory.StartNew(() =>
+            {
+                _documentResolver.TryResolveDocument(csharpParams.RazorFileUri.GetAbsoluteOrUNCPath(), out var documentSnapshot);
+                return documentSnapshot;
+            }, cancellationToken, TaskCreationOptions.None, _foregroundDispatcher.ForegroundScheduler).ConfigureAwait(false);
+
+            if (documentSnapshot is null)
+            {
+                return codeAction;
+            }
+
+            var documentChanged = resolvedCodeAction.Edit.DocumentChanges.First();
+
+            // Only Text Document Edit changes are supported currently
+            if (!documentChanged.IsTextDocumentEdit)
+            {
+                return codeAction;
+            }
+
+            var codeDocument = await documentSnapshot.GetGeneratedOutputAsync().ConfigureAwait(false);
+            var generatedCode = codeDocument.GetCSharpDocument().GeneratedCode;
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var oldText = SourceText.From(generatedCode);
+            var razorDocumentEdits = new List<TextEdit>();
+
+            foreach (var generatedCodeEdit in documentChanged.TextDocumentEdit.Edits)
```

Comment:
Why do this per edit? Why not collect all edits and then apply them all at once?

Reply:
Thanks for pointing this out, ended up being able to remove this and simplify a couple of other areas as well in turn! 🥇